### ISSUE TITLE
Opacus DDP integration

### ISSUE THREAD

**Issue by thomasflynn918** [pytorch/opacus#191]
Hello,

We've been working on DP training with PyTorch's Distributed DataParallel library (DDP) and we thought you might be interested. DDP enables faster training for Opacus over multiple GPUs, compared to the existing implementation for distributed training.

### Overview

The naive implementation for distributed training (e.g., this tutorial: [Writing Distributed Applications with PyTorch — PyTorch Tutorials 1.9.0+cu102 documentation](https://pytorch.org/tutorials/intermediate/dist_tuto.html#distributed-training), or the existing code in Opacus: [Multi-GPU example on CIFAR-10 (#170) · pytorch/opacus@f29110d · GitHub](https://github.com/pytorch/opacus/commit/f29110db3d408e6cf78546f342f7177b2cdb1c28)) works as follows:

- Split the data across nodes
- Do the forward pass on each node
- Do the backward pass on each node
- Clip and add noise to the gradients on each node (with either per-layer clipping or flat clipping)
- Synchronize the noisy gradients by averaging them across all nodes using `torch.distributed`'s all_reduce operator
- Take an optimizer step on each node using the synchronized gradients
  

Instead, our DDP integration with Opacus works as follows:

- Split the data across nodes
- Do the forward pass on each node
- Do the backward pass and the aggregation *in parallel*, thanks to `torch.nn.parallel.DistributedDataParallel`. During this process, the gradients *for each layer* are clipped, noised, and sent to the other nodes for aggregation *as soon as* they become available from backpropagation.  This is implemented using PyTorch's backward hook mechanism.
- Take an optimizer step on each node (on the same gradients)

Logically, our DDP integration with Opacus should be equivalent with distributed Opacus with per-layer clipping.  From a performance perspective, our implementation is much faster than distributed Opacus because it overlaps gradient computations with synchronization.
  
### Basic evaluation
Here is a quick comparison of the performance for our Opacus-DDP implementation vs Opacus' own distributed implementation.  We train a ResNet-50 over Imagenet using 32 GPUs, with a local batch size of 32 images per GPU (for a total batch size of 1024.)  The error bars represent 1 standard deviation:

![](https://i.imgur.com/X4ovgZS.png)

The graph confirms that our Opacus DDP implementation is faster than Opacus' distributed implementation, both with per-layer clipping and with flat clipping (labeled "Naive per-layer clipping" and "Naive flat clipping", respectively).  Our Opacus DDP implementation is similar in performance with that of pure DDP without DP (labeled "DDP (no DP)").  The DDP library paper documents the reason why the DDP library is faster when scaled to multiple GPUs compared to other more naive implementations of distributed-data parallel: [PyTorch Distributed: Experiences on Accelerating Data Parallel Training](https://arxiv.org/abs/2006.15704). Our key insight in this work is to integrate Opacus with this efficient distribution library.

At the moment, DP training on ImageNet does not give a reasonable accuracy-privacy tradeoff.  However, as a sanity check, we verified that after the first five epochs, the test accuracy of our Opacus-DDP implementation is very similar to those of Opacus' distributed implementation with per-layer and flat clipping: 0.43% for all implementations.  All implementations learn something, as the accuracy of the random classifier is 0.1% for this 1000-class task.  However, all configurations learn much less than no-DP SGD does, whose test accuracy after 5 epochs is 27.7%.

The benefit we hypothesize for our scalable Opacus DDP implementation, compared to slower distributed implementations, is that one might obtain a better accuracy-privacy tradeoff for their task by training on a larger dataset but still within reasonable timeframe.  We are currently working on a demonstration of this hypothesis on ImageNet.

### Further notes

- Our Opacus DDP hook is based on the latest release of Opacus (0.13.0), before the GradSample module was introduced, so there is more work to be done if we want to merge it into master.  Moreover, the hook currently supports only per-layer clipping, since it is a hook that runs as soon as a layer is ready. It might be possible to add flat clipping, but at a cost in performance.

- The current multi-GPU implementation ([Multi-GPU example on CIFAR-10 (#170) · pytorch/opacus@f29110d · GitHub](https://github.com/pytorch/opacus/commit/f29110db3d408e6cf78546f342f7177b2cdb1c28)) seems to have a few issues (that you might already be aware of):
  - It doesn't actually run with the latest commits on `master`, because there is a GradSample wrapper around the DPDDP module, which makes [this line](https://github.com/pytorch/opacus/blob/e0b8c676ffbfea9050a46370db3a8b5c56c08e9b/opacus/privacy_engine.py#L148) fail  in the Privacy Engine.
  - The CIFAR example does not seem to use a distributed sampler, so each node is working on the same data instead of splitting the load. We implemented a distributed Poisson sampler to fix this issue in our prototype.
    

Would you be interested in incorporating our DDP hook into Opacus?  We'd be happy to share more details and work with you to integrate the code.

Thanks!

The ExaPPAI (Exascale Privacy-Preserving AI) team at Brookhaven National Laboratory and Columbia University

**Comment by alexandresablayrolles** [comment]
Hi,

This proposal looks great! We are definitely interested in having this in Opacus. Can you submit a PR? We're happy to help you integrate it with Opacus.

**Comment by tholop** [comment]
Thanks @alexandresablayrolles! I opened a PR here: https://github.com/pytorch/opacus/pull/196

**Comment by alexandresablayrolles** [comment]
Closing the issue as PR has been merged.
